File: problems/tier1-foundation/001-payment-ranker/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

// ─── Data Structure ──────────────────────────────────────────────────────────

struct PaymentMethod {
    string name;
    double cashbackRate;        // e.g. 0.05 = 5%
    double transactionFee;      // in rupees
    int    usageCount;
    bool   easyRefundEligible;  // Part 3: refund eligibility flag

    // Default easyRefundEligible to false so Part 1/2 tests
    // that construct with 4 args still compile.
    PaymentMethod(string n, double cb, double fee, int usage, bool refund = false)
        : name(move(n)), cashbackRate(cb), transactionFee(fee),
          usageCount(usage), easyRefundEligible(refund) {}
};

// ─── Strategy Interface ───────────────────────────────────────────────────────

class RankingStrategy {
public:
    virtual bool compare(const PaymentMethod& a, const PaymentMethod& b) = 0;
    virtual ~RankingStrategy() = default;
};

// ─── Part 1: Concrete Strategies ─────────────────────────────────────────────

class RewardsMaximizer : public RankingStrategy {
public:
    bool compare(const PaymentMethod& a, const PaymentMethod& b) override {
        return a.cashbackRate > b.cashbackRate;
    }
};

class LowFeeSeeker : public RankingStrategy {
public:
    bool compare(const PaymentMethod& a, const PaymentMethod& b) override {
        return a.transactionFee < b.transactionFee;
    }
};

class TrustBasedRanker : public RankingStrategy {
public:
    bool compare(const PaymentMethod& a, const PaymentMethod& b) override {
        return a.usageCount > b.usageCount;
    }
};

// ─── Part 2: Composite Strategy ──────────────────────────────────────────────

class CompositeStrategy : public RankingStrategy {
private:
    vector<RankingStrategy*> criteria;
public:
    CompositeStrategy(vector<RankingStrategy*> c) : criteria(move(c)) {}
    bool compare(const PaymentMethod& a, const PaymentMethod& b) override {
        for (auto* s : criteria) {
            if (s->compare(a, b)) return true;   // a wins on this criterion
            if (s->compare(b, a)) return false;   // b wins on this criterion
        }
        return false;  // complete tie across all criteria
    }
};

// ─── Part 3: Easy-Refund Strategy ────────────────────────────────────────────

class EasyRefundStrategy : public RankingStrategy {
private:
    bool prefer;
public:
    EasyRefundStrategy(bool preferRefund) : prefer(preferRefund) {}
    bool compare(const PaymentMethod& a, const PaymentMethod& b) override {
        if (!prefer) return false;
        // a wins if it has easy refund and b does not
        return a.easyRefundEligible && !b.easyRefundEligible;
    }
};

// ─── PaymentRanker (context class) ───────────────────────────────────────────

class PaymentRanker {
private:
    RankingStrategy* strategy;
public:
    PaymentRanker(RankingStrategy* s) : strategy(s) {}
    void setStrategy(RankingStrategy* s) { strategy = s; }
    vector<PaymentMethod> rank(vector<PaymentMethod> methods) {
        sort(methods.begin(), methods.end(),
             [this](const PaymentMethod& a, const PaymentMethod& b) {
                 return strategy->compare(a, b);
             });
        return methods;
    }
};

// ─── Free-function wrappers (used by tests) ──────────────────────────────────

vector<PaymentMethod> rank_by_rewards(vector<PaymentMethod> methods) {
    RewardsMaximizer s;
    return PaymentRanker(&s).rank(methods);
}

vector<PaymentMethod> rank_by_low_fee(vector<PaymentMethod> methods) {
    LowFeeSeeker s;
    return PaymentRanker(&s).rank(methods);
}

vector<PaymentMethod> rank_by_trust(vector<PaymentMethod> methods) {
    TrustBasedRanker s;
    return PaymentRanker(&s).rank(methods);
}

vector<PaymentMethod> rank_composite(vector<PaymentMethod> methods, vector<RankingStrategy*> criteria) {
    CompositeStrategy s(criteria);
    return PaymentRanker(&s).rank(methods);
}

vector<PaymentMethod> rank_with_refund_filter(vector<PaymentMethod> methods, bool preferEasyRefund) {
    EasyRefundStrategy refundStrat(preferEasyRefund);
    RewardsMaximizer rewardsStrat;
    // Refund filter takes priority; cashback is the tiebreaker
    CompositeStrategy s({&refundStrat, &rewardsStrat});
    return PaymentRanker(&s).rank(methods);
}

// ─── Main (guarded so test harness can include this file) ────────────────────

#ifndef RUNNING_TESTS
int main() {
    vector<PaymentMethod> methods = {
        {"HDFC Credit Card", 0.05, 2.0, 120},
        {"UPI",              0.00, 0.0, 450},
        {"Amazon Pay",       0.03, 0.0, 200},
        {"Debit Card",       0.01, 1.0,  80},
    };

    RewardsMaximizer rewardsStrategy;
    PaymentRanker ranker(&rewardsStrategy);

    auto ranked = ranker.rank(methods);
    cout << "Ranked by Rewards:\n";
    for (const auto& m : ranked) {
        cout << "  " << m.name << " (" << m.cashbackRate * 100 << "% cashback)\n";
    }

    cout << "\nRanked with Easy-Refund preference:\n";
    vector<PaymentMethod> refundMethods = {
        {"Card A", 0.10, 5.0, 300, false},
        {"Card B", 0.02, 2.0, 500, true},
        {"Card C", 0.05, 3.0, 400, true},
    };
    auto refundRanked = rank_with_refund_filter(refundMethods, true);
    for (const auto& m : refundRanked) {
        cout << "  " << m.name << " (refund=" << m.easyRefundEligible
             << ", cashback=" << m.cashbackRate * 100 << "%)\n";
    }

    return 0;
}
#endif

File: problems/tier1-foundation/001-payment-ranker/solution.java
// Payment Ranker — Solution (Java)
// All classes in default package; free functions live as static methods on `Solution`.

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

// ─── Data Structure ──────────────────────────────────────────────────────────

class PaymentMethod {
    public String name;
    public double cashbackRate;
    public double transactionFee;
    public int usageCount;
    public boolean easyRefundEligible;

    // 4-arg ctor (Parts 1 & 2): defaults easyRefundEligible to false.
    public PaymentMethod(String name, double cashbackRate, double transactionFee, int usageCount) {
        this(name, cashbackRate, transactionFee, usageCount, false);
    }

    // 5-arg ctor (Part 3).
    public PaymentMethod(String name, double cashbackRate, double transactionFee,
                         int usageCount, boolean easyRefundEligible) {
        this.name = name;
        this.cashbackRate = cashbackRate;
        this.transactionFee = transactionFee;
        this.usageCount = usageCount;
        this.easyRefundEligible = easyRefundEligible;
    }
}

// ─── Strategy Interface ──────────────────────────────────────────────────────

interface RankingStrategy {
    /** Returns true if `a` should come before `b`. */
    boolean compare(PaymentMethod a, PaymentMethod b);
}

// ─── Part 1: Concrete Strategies ─────────────────────────────────────────────

class RewardsMaximizer implements RankingStrategy {
    @Override
    public boolean compare(PaymentMethod a, PaymentMethod b) {
        return a.cashbackRate > b.cashbackRate;
    }
}

class LowFeeSeeker implements RankingStrategy {
    @Override
    public boolean compare(PaymentMethod a, PaymentMethod b) {
        return a.transactionFee < b.transactionFee;
    }
}

class TrustBasedRanker implements RankingStrategy {
    @Override
    public boolean compare(PaymentMethod a, PaymentMethod b) {
        return a.usageCount > b.usageCount;
    }
}

// ─── Part 2: Composite Strategy ──────────────────────────────────────────────

class CompositeStrategy implements RankingStrategy {
    private final List<RankingStrategy> criteria;

    public CompositeStrategy(List<RankingStrategy> criteria) {
        this.criteria = new ArrayList<>(criteria);
    }

    @Override
    public boolean compare(PaymentMethod a, PaymentMethod b) {
        for (RankingStrategy s : criteria) {
            if (s.compare(a, b)) return true;
            if (s.compare(b, a)) return false;
        }
        return false;
    }
}

// ─── Part 3: Easy-Refund Strategy ────────────────────────────────────────────

class EasyRefundStrategy implements RankingStrategy {
    private final boolean prefer;

    public EasyRefundStrategy(boolean preferRefund) {
        this.prefer = preferRefund;
    }

    @Override
    public boolean compare(PaymentMethod a, PaymentMethod b) {
        if (!prefer) return false;
        return a.easyRefundEligible && !b.easyRefundEligible;
    }
}

// ─── PaymentRanker (context class) ───────────────────────────────────────────

class PaymentRanker {
    private RankingStrategy strategy;

    public PaymentRanker(RankingStrategy s) {
        this.strategy = s;
    }

    public void setStrategy(RankingStrategy s) {
        this.strategy = s;
    }

    public List<PaymentMethod> rank(List<PaymentMethod> methods) {
        List<PaymentMethod> out = new ArrayList<>(methods);
        // Convert "a before b" relation into a stable, transitive Comparator.
        out.sort((a, b) -> {
            if (strategy.compare(a, b)) return -1;
            if (strategy.compare(b, a)) return 1;
            return 0;
        });
        return out;
    }
}

// ─── Free-function wrappers (used by tests) ──────────────────────────────────

public class Solution {
    public static List<PaymentMethod> rank_by_rewards(List<PaymentMethod> methods) {
        return new PaymentRanker(new RewardsMaximizer()).rank(methods);
    }

    public static List<PaymentMethod> rank_by_low_fee(List<PaymentMethod> methods) {
        return new PaymentRanker(new LowFeeSeeker()).rank(methods);
    }

    public static List<PaymentMethod> rank_by_trust(List<PaymentMethod> methods) {
        return new PaymentRanker(new TrustBasedRanker()).rank(methods);
    }

    public static List<PaymentMethod> rank_composite(
            List<PaymentMethod> methods, List<RankingStrategy> criteria) {
        return new PaymentRanker(new CompositeStrategy(criteria)).rank(methods);
    }

    public static List<PaymentMethod> rank_with_refund_filter(
            List<PaymentMethod> methods, boolean preferEasyRefund) {
        List<RankingStrategy> chain = new ArrayList<>();
        chain.add(new EasyRefundStrategy(preferEasyRefund));
        chain.add(new RewardsMaximizer());
        return new PaymentRanker(new CompositeStrategy(chain)).rank(methods);
    }
}
